File: src/structs/grid.js
import { CATEGORY, PARTICLE, PARTICLE_DATA } from "../structs/data.js";
import { Color } from "../structs/color.js";
import { Utility } from "../utils/utility.js";

export class Grid {
  width;
  height;
  #data;
  #dirtyParticles;
  NEIGHBOR = {
    TOP_LEFT: { dx: -1, dy: -1 },
    TOP_MIDDLE: { dx: 0, dy: -1 },
    TOP_RIGHT: { dx: 1, dy: -1 },
    LEFT: { dx: -1, dy: 0 },
    RIGHT: { dx: 1, dy: 0 },
    DOWN_LEFT: { dx: -1, dy: 1 },
    DOWN_MIDDLE: { dx: 0, dy: 1 },
    DOWN_RIGHT: { dx: 1, dy: 1 },
  };
  PROCESSED_PARTICLE_DATA = {};

  constructor(width, height, dataOverride = null) {
    this.width = width;
    this.height = height;
    this.#data = new Array(width * height).fill(null);
    this.#dirtyParticles = new Set();

    this.#processParticleData(PARTICLE_DATA);
  }
  get data() {
    return this.#data;
  }
  get dirtyParticles() {
    return this.#dirtyParticles;
  }
  isInBounds(x, y) {
    return x >= 0 && x < this.width && y >= 0 && y < this.height;
  }
  clearDirty() {
    this.#dirtyParticles.clear();
  }

  // --------- Helper Functions ---------
  #processParticleData(rawParticleData) {
    for (const key in rawParticleData) {
      const particle = rawParticleData[key];

      // Modify particle data
      if (particle.CATEGORY === CATEGORY.SAND) {
        particle.REPOSE_DIRECTIONS = Utility.calculateRepose(particle.REPOSE_ANGLE);
      }

      this.PROCESSED_PARTICLE_DATA[key] = particle;
    }

    // Make the newly created particle data a const
    Object.freeze(this.PROCESSED_PARTICLE_DATA);
  }

  // Particle factory
  #createNewParticle(particleId) {
    const particleData = this.PROCESSED_PARTICLE_DATA[particleId];

    // Generate a random color for this particle between the particle's base and variant colors
    const baseColor = Color.hexToRGBA(particleData.COLOR_BASE);
    const variantColor = Color.hexToRGBA(particleData.COLOR_VARIANT);
    const newColor = Color.colorBetween(baseColor, variantColor, Math.random());

    let newParticle = {
      id: particleData.ID,
      name: particleData.NAME,
      color: newColor,
      position: null,
      index: 0,
      category: particleData.CATEGORY,
      isMovable: particleData.ISMOVABLE,
      density: particleData.DENSITY,
    };

    // ..Then we assign category specific properties to it...
    switch (particleData.CATEGORY) {
      case CATEGORY.SOLID:
        break;
      case CATEGORY.LIQUID:
        newParticle.concentration = 1;
        newParticle.maxConcentration = particleData.MAXCONCENTRATION;
        break;
      case CATEGORY.GAS:
        break;
      case CATEGORY.SAND:
        newParticle.reposeAngle = particleData.REPOSE_ANGLE;
        newParticle.reposeDirections = particleData.REPOSE_DIRECTIONS;
        break;
      case CATEGORY.ELECTRONICS:
        newParticle.node = null;
        break;
      default:
        console.log("Error: None implemented category type:" + newParticle);
    }

    return newParticle;
  }

  // --------- Public Functions ---------

  // Create a new particle at the given x and y coords. Existing particle at that coords will be removed.
  createParticleAt(x, y, particleId, markAsDirty = false, markNeighborsAsDirty = false) {
    // Don't create a particle if the given location and the particle id is invalid
    if (!this.isInBounds(x, y) || !particleId) return false;

    // Create a new particle and assign it's position in the grid
    let newParticle = this.#createNewParticle(particleId);
    newParticle.position = { x: x, y: y };
    newParticle.index = y * this.width + x;

    // Add the newly created particle to the grid
    this.#data[y * this.width + x] = newParticle;

    // Handle dirty particles
    if (markAsDirty) {
      // Add the neighboring particles too
      if (markNeighborsAsDirty) {
        const newParticleNeighbors = this.getValidNeighborParticles(newParticle, this.NEIGHBOR);
        if (newParticleNeighbors) {
          for (const neighbor of newParticleNeighbors) this.#dirtyParticles.add(neighbor);
        }
      }

      this.#dirtyParticles.add(newParticle);
    }
  }

  // Returns particle reference at x and y coords, returns null if coords are out of bounds
  getParticleAt(x, y) {
    if (!this.isInBounds(x, y)) return null;
    return this.#data[y * this.width + x];
  }

  // Tries to move a particle based on provided direction groups. Stops as soon as the particle moves.
  tryMoveParticle(particle, directionGroups, dumpThemNerds = false, markAsDirty = false, markNeighborsAsDirty = false) {
    if (!particle || !directionGroups) return null;

    for (const directions of directionGroups) {
      if (!directions) continue;
      // Shuffle the directions for random movements
      const shuffledDirections = directions.length > 1 ? Utility.shuffleArray(directions) : directions;

      // Iterate through directions and check if particle can move there
      for (const direction of shuffledDirections) {
        // Add random 'dumps' in the x axis
        let finalDX = direction.dx;
        if (dumpThemNerds) {
          if (Math.random() > 0.5) {
            finalDX = direction.dx * -2;
          }
        }

        // Get the target particle
        const targetX = particle.position.x + finalDX;
        const targetY = particle.position.y + direction.dy;

        const target = this.getParticleAt(targetX, targetY);
        if (!target) continue;

        // If target particle is movable and has lower density than the current particle..
        // ..We move the current particle to the target particle (swap them)
        if (target.isMovable && particle.density > target.density) {
          // Swap particles in the data array
          this.#data[target.index] = particle;
          this.#data[particle.index] = target; // particle's index hasn't changed yet, so we can safly use it here

          // Update their postion and index
          const particlePosition = { x: particle.position.x, y: particle.position.y };
          const particleIndex = particle.index;
          particle.position = target.position;
          target.position = particlePosition;
          particle.index = target.index;
          target.index = particleIndex;

          // Mark them as dirty
          if (markAsDirty) {
            this.#dirtyParticles.add(particle);
            this.#dirtyParticles.add(target);

            // Add the neighbors of the NEW positions to the dirty set
            if (markNeighborsAsDirty) {
              const particleNeighbors = this.getValidNeighborParticles(particle, this.NEIGHBOR);
              const targetNeighbors = this.getValidNeighborParticles(target, this.NEIGHBOR);
              if (particleNeighbors) {
                for (const neighbor of particleNeighbors) {
                  this.#dirtyParticles.add(neighbor);
                }
              }
              if (targetNeighbors) {
                for (const neighbor of targetNeighbors) {
                  this.#dirtyParticles.add(neighbor);
                }
              }
            }
          }

          return target; // Particle was moved successfully
        }
      }
    }
    return null; // Particle could not be moved successfully
  }

  // Swap particle A and B location on the data array.
  swapParticles(particleAIndex, particleBIndex, markAsDirty = false, markNeighborsAsDirty = false) {
    if (particleAIndex === particleBIndex) return false;
    if (particleAIndex < 0 || particleAIndex >= this.#data.length) return false;
    if (particleBIndex < 0 || particleBIndex >= this.#data.length) return false;

    const particleA = this.#data[particleAIndex];
    const particleB = this.#data[particleBIndex];
    if (!particleA || !particleB) return false;

    const particleAPos = { ...particleA.position };
    const particleBPos = { ...particleB.position };

    particleA.position = particleBPos;
    particleB.position = particleAPos;
    this.#data[particleAIndex] = particleB;
    this.#data[particleBIndex] = particleA;

    // Recalculate indexes for both particles
    particleA.index = particleA.position.y * this.width + particleA.position.x;
    particleB.index = particleB.position.y * this.width + particleB.position.x;

    // Dirty cell handling
    if (markAsDirty) {
      // Add the new positions of both particles to the dirty set
      this.#dirtyParticles.add(particleA);
      this.#dirtyParticles.add(particleB);

      // Add the neighbors of the NEW positions to the dirty set
      if (markNeighborsAsDirty) {
        const particleANeighbors = this.getValidNeighborParticles(particleA, this.NEIGHBOR);
        const particleBNeighbors = this.getValidNeighborParticles(particleB, this.NEIGHBOR);
        if (particleANeighbors) {
          for (const neighbor of particleANeighbors) {
            this.#dirtyParticles.add(neighbor);
          }
        }
        if (particleBNeighbors) {
          for (const neighbor of particleBNeighbors) {
            this.#dirtyParticles.add(neighbor);
          }
        }
      }
    }

    return true;
  }

  // Returns a neighbor in the offset direction of a given particle
  getNeighborParticle(particle, offset) {
    const neighborX = particle.position.x + offset.dx;
    const neighborY = particle.position.y + offset.dy;

    return this.getParticleAt(neighborX, neighborY);
  }

  // Always returns an array that's either empty or has valid neighbors
  getValidNeighborParticles(particle, offsets, category = CATEGORY.ANY, type = PARTICLE.ANY) {
    let neighbors = [];
    if (!particle) return neighbors;

    for (const key in offsets) {
      const offset = offsets[key];
      const neighborX = particle.position.x + offset.dx;
      const neighborY = particle.position.y + offset.dy;
      const neighbor = this.getParticleAt(neighborX, neighborY);
      if (!neighbor) continue; // if neighbor is null or invalid, don't bother

      // Category and type checks
      const isCategoryTrue = category === CATEGORY.ANY || neighbor.category === category;
      const isTypeTrue = type === PARTICLE.ANY || neighbor.id === type;

      if (isCategoryTrue && isTypeTrue) {
        neighbors.push(neighbor);
      }
    }
    return neighbors;
  }

  //
  populateGrid(particleId) {
    const gridWidth = this.width;
    const gridHeight = this.height;

    for (let y = 0; y < gridHeight; y++) {
      for (let x = 0; x < gridWidth; x++) {
        this.createParticleAt(x, y, particleId, true);
      }
    }
  }

  //
  markParticleDirty(particle, markNeighborsAsDirty = false) {
    this.#dirtyParticles.add(particle);

    // Add the neighbors of the NEW positions to the dirty set
    if (markNeighborsAsDirty) {
      const particleNeighbors = this.getValidNeighborParticles(particle, this.NEIGHBOR);
      if (particleNeighbors) {
        for (const neighbor of particleNeighbors) {
          this.#dirtyParticles.add(neighbor);
        }
      }
    }
  }

  // Function to draw a circle of particles in the grid
  fillCircleAt(x, y, radius, particleId, concentrationOverride = 1) {
    const newConcentration = concentrationOverride > 0 ? concentrationOverride : 1;

    for (let i = -radius; i <= radius; i++) {
      for (let j = -radius; j <= radius; j++) {
        if (i * i + j * j <= radius * radius) {
          const px = x + i;
          const py = y + j;

          // Only draw if the particle to draw is within current grid's bounds
          if (px >= 0 && px < this.width && py >= 0 && py < this.height) {
            const prevParticle = this.getParticleAt(px, py);

            if (prevParticle === null || prevParticle.id === PARTICLE.EMPTY || particleId === PARTICLE.EMPTY) {
              this.createParticleAt(px, py, particleId, true, true);
              this.getParticleAt(px, py).concentration = newConcentration;
            }
          }
        }
      }
    }
  }
}

File: src/structs/grid.rs
use crate::loader::load_particle_data;
use crate::structs::particle::Particle;
use crate::structs::particle_data::ParticleData;
use crate::structs::utils::{Offset2, Vector2};
use std::collections::HashMap;
use std::collections::HashSet;

use rand::{Rng, prelude::SliceRandom, prelude::ThreadRng, random};

const MOORE_NEIGHBORS: [(i32, i32); 8] = [(-1, -1), (0, -1), (1, -1), (-1, 0), (1, 0), (-1, 1), (0, 1), (1, 1)];
const VON_NEUMANN_NEIGHBORS: [(i32, i32); 4] = [(0, -1), (-1, 0), (1, 0), (0, 1)];

pub struct Grid {
    pub width: i32,
    pub height: i32,
    pub data: Vec<Particle>,
    pub dirty_particles: HashSet<u32>,

    particle_data_map: HashMap<u16, ParticleData>,
}

impl Grid {
    pub fn new(width: u16, height: u16) -> Grid {
        // Load particle data for this grid
        let particle_data_map: HashMap<u16, ParticleData> = load_particle_data("./src/data/particles.data");

        // Create a new empty data for this grid
        let data: Vec<Particle> = Vec::with_capacity(width as usize * height as usize);

        return Grid {
            width: width as i32,
            height: height as i32,
            data: data,
            dirty_particles: HashSet::<u32>::new(),

            particle_data_map: particle_data_map,
        };
    }

    // --------- Helper Functions ---------
    pub fn is_in_bounds(&self, x: i32, y: i32) -> bool {
        return x >= 0 && x < self.width && y >= 0 && y < self.height;
    }

    pub fn populate(&mut self, particle_id: u16) {
        // Retrive particle from particle data map using the provided id
        let particle_data: &ParticleData = self.particle_data_map.get(&particle_id).unwrap();

        // Create grid data and populate it with the provided particles
        for y in 0..self.height {
            for x in 0..self.width {
                let index: u32 = (y * self.width + x) as u32;
                let handle: u32 = index;

                let mut particle: Particle = Particle::new(particle_data, x, y, handle);
                particle.index = index;
                self.data.push(particle);
            }
        }
    }

    pub fn mark_particle_dirty(&mut self, x: i32, y: i32, mark_neighbors_dirty: bool) {
        // Get particle's index
        let particle_index: u32 = (y * self.width + x) as u32;

        // Add the particle's index to mark it dirty
        self.dirty_particles.insert(particle_index);

        // Mark particle's neighbors dirty
        if mark_neighbors_dirty {
            // Get particle's neighbors
            let neighbor_indices: Vec<u32> = self.get_neighbor_indices_of(x, y, &MOORE_NEIGHBORS);

            // Mark them dirty
            for neighbor_index in neighbor_indices {
                self.dirty_particles.insert(neighbor_index);
            }
        }
    }

    // --------- Methods ---------

    pub fn create_particle_at(&mut self, x: i32, y: i32, particle_id: u16, mark_dirty: bool, mark_neighbors_dirty: bool) -> bool {
        // x and y axis are out of bounds, return
        if !(self.is_in_bounds(x, y)) {
            return false;
        }

        // Retrieve particle data
        let particle_data: &ParticleData = match self.particle_data_map.get(&particle_id) {
            Some(val) => val,
            None => return false, // Cannot create particle, invalid particle id, return
        };

        // Instanciate a new particle
        let mut new_particle: Particle = Particle::new(particle_data, x, y, 1);

        // Set the new particle's index
        let index: u32 = (y * self.width + x) as u32;
        new_particle.index = index;

        // Assigning the new particle to the data
        self.data[index as usize] = new_particle;

        // Handle dirty logic
        if mark_dirty {
            self.mark_particle_dirty(x, y, mark_neighbors_dirty);
        }

        return true;
    }

    pub fn get_particle_at<'a>(&'a self, x: i32, y: i32) -> Option<&'a Particle> {
        // Check if the position is out of bounds
        if !(self.is_in_bounds(x, y)) {
            return None;
        }

        let index = y * self.width + x;
        return self.data.get(index as usize);
    }

    pub fn get_neighbor_indices_of(&self, x: i32, y: i32, offsets: &[(i32, i32)]) -> Vec<u32> {
        let mut neighbors: Vec<u32> = Vec::with_capacity(offsets.len());

        for (x_offset, y_offset) in offsets {
            // Convert offset position to world position
            let x_pos: i32 = x + x_offset;
            let y_pos: i32 = y + y_offset;

            // Add neighbor to the list if it's in grid's bounds
            if self.is_in_bounds(x_pos, y_pos) {
                let neighbor_index: u32 = (y_pos * self.width + x_pos) as u32;
                neighbors.push(neighbor_index);
            }
        }

        return neighbors;
    }
    pub fn try_move_particle(
        &mut self,
        particle_index: usize,
        direction_groups: &[Vec<Offset2<i32>>],
        bump_them_nerds: bool,
        mark_dirty: bool,
        mark_neighbors_dirty: bool,
    ) -> bool {
        let current_position = self.data[particle_index].position;
        let current_density = self.data[particle_index].density;
        let width = self.width;

        for directions in direction_groups {
            for direction in directions {
                // Add random 'bumps' in the x axis
                let mut final_dx: i32 = direction.dx;
                if bump_them_nerds {
                    if rand::random::<f32>() > 0.5 {
                        final_dx = -direction.dx;
                    }
                }

                let tx: i32 = current_position.x + final_dx;
                let ty: i32 = current_position.y + direction.dy;

                if !self.is_in_bounds(tx, ty) {
                    continue;
                }

                let target_index: usize = (ty * width + tx) as usize;

                if particle_index == target_index {
                    continue;
                }

                let moved: bool = {
                    let (a, b) = if particle_index < target_index {
                        let (a, b) = self.data.split_at_mut(target_index);
                        (&mut a[particle_index], &mut b[0])
                    } else {
                        let (a, b) = self.data.split_at_mut(particle_index);
                        (&mut b[0], &mut a[target_index])
                    };

                    if b.is_movable && current_density > b.density {
                        std::mem::swap(a, b);
                        let temp_pos = a.position;
                        a.position = b.position;
                        b.position = temp_pos;

                        let temp_index = a.index;
                        a.index = b.index;
                        b.index = temp_index;

                        true
                    } else {
                        false
                    }
                };

                if moved {
                    if mark_dirty {
                        self.mark_particle_dirty(tx, ty, mark_neighbors_dirty);
                        self.mark_particle_dirty(current_position.x, current_position.y, mark_neighbors_dirty);
                    }
                    return true;
                }
            }
        }

        return false;
    }

    pub fn fill_circle_at(&mut self, x: i32, y: i32, radius: i32, particle_id: u16) {
        for i in -radius..radius {
            for j in -radius..radius {
                if i * i + j * j <= radius * radius {
                    let px: i32 = x + i;
                    let py: i32 = y + j;

                    // Particle to draw is out of grid's bounds, skip it
                    if !self.is_in_bounds(px, py) {
                        continue;
                    }

                    // Create particle
                    self.create_particle_at(px, py, particle_id, true, true);
                }
            }
        }
    }
}
